src/main/scala/org/hammerlab/guacamole/loci/partitioning/ApproximatePartitioner.scala

```diff
@@ -0,0 +1,190 @@
+import org.hammerlab.guacamole.loci.map.LociMap
+import org.hammerlab.guacamole.loci.partitioning.ApproximatePartitioner.{MicroPartitionIndex, NumMicroPartitions}
+import org.hammerlab.guacamole.loci.partitioning.LociPartitioner.{NumPartitions, PartitionIndex}
+import org.hammerlab.guacamole.loci.set.LociSet
+import org.hammerlab.guacamole.logging.LoggingUtils.progress
+import org.hammerlab.guacamole.reference.ReferenceRegion
+import org.kohsuke.args4j.{Option => Args4jOption}
+
+import scala.collection.Map
+import scala.reflect.ClassTag
+
+trait ApproximatePartitionerArgs extends UniformPartitionerArgs {
+
+  /**
+   * Long >= 1. Number of micro-partitions generated for each of the `numPartitions` Spark partitions that will be
+   * computed. Higher values of this will result in a more exact but more expensive computation.
+   * In the extreme, setting this to greater than the number of loci (per partition) will result in an exact
+   * calculation.
+   */
+  @Args4jOption(
+    name = "--partition-accuracy",
+    usage = "Number of micro-partitions to use, per Spark partition, when partitioning loci (default: 250). Set to 0 to partition loci uniformly"
+  )
+  var partitioningAccuracy: NumMicroPartitions = 250
+}
+
+/**
+ * Assign loci from a LociSet to partitions, where each partition overlaps approximately the same number of "regions"
+ * (reads mapped to a reference genome).
+ *
+ * The approach we take is:
+ *
+ *  (1) chop up the loci uniformly into many genomic "micro partitions."
+ *
+ *  (2) for each micro partition, calculate the number of regions that overlap it.
+ *
+ *  (3) using these counts, assign loci to real (Spark) partitions, assuming approximately uniform depth within each
+ *      micro partition.
+ *
+ *  Some advantages of this approach are:
+ *
+ *  - Stages (1) and (3), which are done locally by the Spark master, are constant time with respect to the number
+ *    of regions (though linear in the number of micro-partitions).
+ *
+ *  - Stage (2), where runtime does depend on the number of regions, is done in parallel with Spark.
+ *
+ *  - Accuracy vs. performance can be tuned by setting `accuracy`.
+ *
+ *  - Does not require a distributed sort.
+ *
+ * @param regions: RDD of reads to base the partitioning on.
+ * @param numPartitions Number of partitions to split reads into.
+ * @param microPartitionsPerPartition Long >= 1. Number of micro-partitions generated for each of the `numPartitions`
+ *                                    Spark partitions that will be computed. Higher values of this will result in a
+ *                                    more exact but more expensive computation.
+ *                                    In the extreme, setting this to greater than the number of loci (per partition)
+ *                                    will result in an exact calculation.
+ * @return LociMap of locus -> partition assignments.
+ */
+class ApproximatePartitioner[R <: ReferenceRegion: ClassTag](regions: RDD[R],
```

Comment:
I'm not sure if `ApproximatePartitioner` is the right name for this. Maybe `ApproximateRegionDepthPartitioner`? Then your new partitioner would be something like `ApproximateRegionDepthWithThresholdPartitioner`. Not that these huge names are that lovely, but I don't think Approximate vs. Exact is really the distinction between this partitioner and yours, unless I'm misunderstanding yours.